### PR TITLE
VSHA-503 modify tests for use on vshasta

### PR DESCRIPTION
**Pull request by jacobsalmela**
Signed-off-by: Jacob Salmela <jacob.salmela@hpe.com>

## Summary and Scope

Add vshasta support to some existing tests by adding a conditional.

Adds a `is_vshasta_node` function.

## Issues and Related PRs

* Partially resolves VSHA-503
* Works with https://github.com/Cray-HPE/livecd-gcp-infrastructure/pull/174

### Tested on:

  * `#redbull`
  * Virtual Shasta

### Test description:

Executed the tests on a vshasta environment and saw the tests were skipped.

The function works on a vshasta pit node:

```
pit:~ # is_vshasta_node ; echo $?
+ [[ -f /etc/google_system ]]
+ return 0
+ echo 0
0
pit:~ #
```

a metal image booted to vshasta:

```
ncn-m002:~ # is_vshasta_node ; echo $?
+ [[ -f /etc/google_system ]]
+ grep --color=auto Google
+ lsblk --noheadings -o vendor
+ return 0
+ echo 0
0
```

and a metal node:

```
redbull-ncn-m001-pit:~ # is_vshasta_node ; echo $?
+ [[ -f /etc/google_system ]]
+ grep Google
+ lsblk --noheadings -o vendor
+ return 1
+ echo 1
1
```

## Risks and Mitigations

Low

## Pull Request Checklist

- [ ] Version number(s) incremented, if applicable
- [ ] Copyrights updated
- [ ] License file intact
- [ ] Target branch correct
- [ ] CHANGELOG.md updated
- [ ] Testing is appropriate and complete, if applicable
- [ ] [HPC Product Announcement](https://cray.slack.com/archives/C026TVCSXLH) prepared, if applicable

### CODE REVIEW

**Review by heemstra — APPROVED** [review]

File: goss-testing/scripts/python/check_for_unused_drives.py
Comment (mharding-hpe) [inline goss-testing/scripts/python/check_for_unused_drives.py:80]: I don't see any other place which would ever set the hardware type to vshasta. Don't you need to modify that code as well?

Also, if this is done, isn't the vshasta variable then redundant? The hardware type could just be checked instead, it seems like, unless I'm missing something here.

File: goss-testing/vars/variables-common.yaml
Comment (mharding-hpe) [inline goss-testing/vars/variables-common.yaml:69]: I don't see any code which is doing this overriding in this PR.

File: start-goss-servers.sh
Comment (mharding-hpe) [inline start-goss-servers.sh:63]: If that command is going to be the general way to check if this is vshasta or not, then I'd recommend adding it to the shared Bash function library (`automated/run-ncn-tests.sh`) as a function like `is_vshasta`, similar to the `is_pit_node` function. That would make it easier to re-use, and if we ever need to change it, we won't have to hunt down every place we use the command.

**Review by mharding-hpe — CHANGES_REQUESTED** [review]

File: goss-testing/scripts/python/check_for_unused_drives.py
Comment (jacobsalmela) [reply goss-testing/scripts/python/check_for_unused_drives.py:80]: There actually is no ipmitool on vshasta.  We just run a mock version of it, so we just fake it.

https://github.com/Cray-HPE/livecd-gcp-infrastructure/pull/174/files#diff-34e0270d5867cf8e17dc52ef091290e1cf3b721c71c3d9181004129abf8aa7b5R207

File: goss-testing/vars/variables-common.yaml
Comment (jacobsalmela) [reply goss-testing/vars/variables-common.yaml:69]: This var went in in a few spots.  It is false by default in the vars files.  There is a hacky override when we run in vshasta to set it to true.  It is false by default so it does nothing on metal installs

**Comment by mharding-hpe** [comment]
In addition to creating the `is_vshasta` function in [the common function library](https://github.com/Cray-HPE/csm-testing/blob/main/goss-testing/automated/run-ncn-tests.sh), it seems like you'd also want to add code to the `add_local_vars` function in that library to handle any new variables you want to have set.

File: start-goss-servers.sh
Comment (jacobsalmela) [reply start-goss-servers.sh:63]: We actually have a better `is_vshasta` check, which is to see if `/etc/google_system` exists, but we boot the metal images, which doesn't have that file, so the method of determining if a node is vshasta or not needs some more consideration I think, rather than what I have here.

File: goss-testing/vars/variables-common.yaml
Comment (jacobsalmela) [reply goss-testing/vars/variables-common.yaml:69]: See https://github.com/Cray-HPE/csm-testing/pull/410/files#diff-fd12da3b417be82600b83d8504e8a7a264a13585ce145769ac363277945447ffR102

Comment (jacobsalmela) [reply goss-testing/vars/variables-common.yaml:69]: All of the overrides happen in the vshasta codebase.

File: start-goss-servers.sh
Comment (mharding-hpe) [reply start-goss-servers.sh:63]: Either way, with whatever method you arrive at, it seems like we'd want it to be in a common function for easier reuse, right?

File: goss-testing/vars/variables-common.yaml
Comment (mharding-hpe) [reply goss-testing/vars/variables-common.yaml:69]: Why would you not put it in here, if we already have a method for determining that this is a Vshasta system? Seems easier to maintain that way, rather than requiring PRs in two different repos if wanting to make a change to that part of it. No?

File: start-goss-servers.sh
Comment (jacobsalmela) [reply start-goss-servers.sh:63]: Yeah, it is not a hard change to make.  I am fixing it up.

File: goss-testing/vars/variables-common.yaml
Comment (jacobsalmela) [reply goss-testing/vars/variables-common.yaml:69]: We certainly could put it in here.  This repo is one of the least developer-friendly repos to contribute to, so that is the main reason I never even considered it.   Also, I already needed to contribute to here, to the csm repo and to csm-rpms in order for my change to show up a few days later, so keeping the vshasta stuff in the vshasta repo meant I could develop it faster. 

I can certainly add it here if that is a gate for you.

Comment (mharding-hpe) [reply goss-testing/vars/variables-common.yaml:69]: I would strongly prefer that we have it here. As you've already seen, this repo contains some tests which are not part of any test suite and may or may not still be in use. The same is true of parts of tests and scripts in here -- there are sections which may or may not still be needed. Down the road, if someone notices that we have a lot of vshasta checks, but that there is no place where we ever set it to true, they may conclude that this is another case of something that is no longer in use.

Also, if a change is made to csm-testing that results in the variable files having a different format, or location, or whatever, right now someone making that change would have no way to know that a change would also be required in the vshasta repo. That is another one of the benefits of doing this in the add_local_vars function. It will automatically be kept up to date.

Put another way, I think it's not ideal to have one repo (your vshasta repo) hard-code in assumptions about how another repo (csm-testing) has its stuff internally organized. The automated scripts we provide are really the only interface that should be relied on externally. For example, we don't have anything outside of the repo which directly calls the Goss endpoints, for exactly this reason. There's no promise made that the URLs, port numbers, etc, won't change.

Comment (jacobsalmela) [reply goss-testing/vars/variables-common.yaml:69]: Makes sense.  I will make the adjustments

Comment (mharding-hpe) [reply goss-testing/vars/variables-common.yaml:69]: For what it's worth, I completely agree with you about this repository not being developer-friendly. That is something we need to improve on.

File: goss-testing/scripts/python/check_for_unused_drives.py
Comment (jacobsalmela) [reply goss-testing/scripts/python/check_for_unused_drives.py:80]: I think we may still want this code.  I think this test should still run on vshasta, rather then being skipped on the `is_vshasta` conditional.

File: goss-testing/automated/run-ncn-tests.sh
Comment (mharding-hpe) [inline goss-testing/automated/run-ncn-tests.sh:51]: Should be the same, but seems more straightforward to not use the variable when we always know it is 0?

```suggestion
      return 0
```

File: goss-testing/automated/run-ncn-tests.sh
Comment (mharding-hpe) [inline goss-testing/automated/run-ncn-tests.sh:54]: Should be the same, but seems more straightforward to not use the variable when we always know it is 0?

```suggestion
      return 0
```

File: goss-testing/scripts/python/check_for_unused_drives.py
Comment (mharding-hpe) [reply goss-testing/scripts/python/check_for_unused_drives.py:80]: Yeah, I agree. I was saying that if you have the hardware manufacturer variable set to `vshasta`, then you don't also need the other `vshasta` boolean variable. You could just check the manufacturer variable. But it's not a big deal.

File: start-goss-servers.sh
Comment (mharding-hpe) [inline start-goss-servers.sh:83]: This will mean that the other scripts that call the endpoints won't work with vshasta, since they expect it to be in json format. Not sure if that is of concern.

File: start-goss-servers.sh
Comment (mharding-hpe) [inline start-goss-servers.sh:69]:
```suggestion
if is_vshasta_node; then
    echo "vshasta detected."

    # Use junit for vshasta
    results_format=junit

    # on vshasta there is only one network and no hmn
    ip=$(host "$(hostname)" | grep -Po '[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}')
else
    # for security reasons we only want to run the servers on the HMN network, which is not connected to open Internet
    ip=$(host "$(hostname).hmn" | grep -Po '[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}')
    
    # Return results in JSON format outside of vshasta
    results_format=json
fi
```

File: start-goss-servers.sh
Comment (mharding-hpe) [inline start-goss-servers.sh:96]:
```suggestion
    # Start Goss server for this entry.
    echo "starting ${endpoint} in background on port ${port}"
    /usr/bin/goss -g "${suite}" --vars "${tmpvars}" serve \
        --format "${results_format}" \
        --max-concurrent 4 \
        --endpoint "/${endpoint}" \
        --listen-addr "${ip}:${port}" &
```

**Review by mharding-hpe — CHANGES_REQUESTED** [review]

File: start-goss-servers.sh
Comment (jacobsalmela) [reply start-goss-servers.sh:83]: I do not think so... as it is now, our deployment is the only thing that runs the tests.

File: goss-testing/scripts/python/check_for_unused_drives.py
Comment (jacobsalmela) [reply goss-testing/scripts/python/check_for_unused_drives.py:80]: True.  It is redundant, but it is a very small component of vshasta and mostly superfluous because we just use it to fake out some other scripts that depend on it, but we actually have the data elsewhere

File: goss-testing/vars/variables-common.yaml
Comment (jacobsalmela) [reply goss-testing/vars/variables-common.yaml:69]: testing after the recent changes

Comment (jacobsalmela) [reply goss-testing/vars/variables-common.yaml:69]: It worked

File: start-goss-servers.sh
Comment (jacobsalmela) [reply start-goss-servers.sh]: Fixed.  More relevant info here: https://github.com/Cray-HPE/csm-testing/pull/414#discussion_r1018211533

File: goss-testing/automated/run-ncn-tests.sh
Comment (mharding-hpe) [reply goss-testing/automated/run-ncn-tests.sh:54]: This was marked resolved but it looks to me like it is not? Or is github messing with me again? Sometimes it likes to show me stale PR data, I think.

File: goss-testing/vars/variables-common.yaml
Comment (mharding-hpe) [reply goss-testing/vars/variables-common.yaml:69]: The code to set this in the `add_local_vars` function should still be added, using the `is_vshasta_node` function that you created.

**Review by mharding-hpe — CHANGES_REQUESTED** [review]
Some of the comments were marked resolved but do not appear to be. I've marked them unresolved and added comments.

File: goss-testing/automated/run-ncn-tests.sh
Comment (jacobsalmela) [reply goss-testing/automated/run-ncn-tests.sh:54]: Oh yeah, I co-authored the commit with you.  https://github.com/Cray-HPE/csm-testing/pull/414/commits/be2345b6a3bea4ce0080b75151b5e46be7d259e0.

The same thing happens to me, but I took all your suggestions if you check the commits tab.

File: goss-testing/vars/variables-common.yaml
Comment (jacobsalmela) [reply goss-testing/vars/variables-common.yaml:69]: oops. I'll look

Comment (jacobsalmela) [reply goss-testing/vars/variables-common.yaml:69]: ok check now

File: goss-testing/automated/run-ncn-tests.sh
Comment (mharding-hpe) [reply goss-testing/automated/run-ncn-tests.sh:54]: I don't see the above suggestion in the commit tag.

**Review by mharding-hpe — CHANGES_REQUESTED** [review]
I am still seeing the return $? on line 54 of `goss-testing/automated/run-ncn-tests.sh`, even if I view it directly from a git clone of the branch.

I also still see a change needed in the `add_local_vars` function in that same file.

To save time, I am going to make a PR against your PR branch with the changes I still think should be made. You can merge it into your PR branch if you agree.

File: goss-testing/automated/run-ncn-tests.sh
Comment (jacobsalmela) [reply goss-testing/automated/run-ncn-tests.sh]: https://github.com/Cray-HPE/csm-testing/pull/414/commits/afc2c42471a264a846a023b26ed51e9fddb9954e

**Review by mharding-hpe — CHANGES_REQUESTED** [review]
Please take a look at https://github.com/Cray-HPE/csm-testing/pull/416

That addresses my remaining concerns,

If you don't have authority to merge it into this branch, just review it and approve it, and I can merge it in here.

File: goss-testing/vars/variables-common.yaml
Comment (mharding-hpe) [reply goss-testing/vars/variables-common.yaml:69]: I see the code for the vshasta variable in there, but it looks like it still sets the node manufacturer using ipmitool, which I believe you said will not work for vshasta.

I made this PR against your PR branch:
https://github.com/Cray-HPE/csm-testing/pull/416

It will address that issue, and also make some small formatting changes to make the new code more closely match the style of the existing functions.

**Comment by jacobsalmela** [comment]
Pulled it in

**Review by mharding-hpe — APPROVED** [review]

**Comment by mharding-hpe** [comment]
> Pulled it in
Perfect. I've approved, but looks like you may need to squash the commits now because of the merge commit that snuck in there.

**Comment by jacobsalmela** [comment]
Removed the merge commit.  I think we're good now

**Comment by mharding-hpe** [comment]
> Removed the merge commit. I think we're good now

Cool. I think you need someone on PET team to approve this before it can merge. Or anyway, I don't have authority to do it without that. Maybe go hit up #casm-pet channel and beg. :)

**Comment by jacobsalmela** [comment]
Looks like I cannot merge it yet either

**Review by bklei — APPROVED** [review]

**Pull request merged**
